src/dynamic_programming/divide-and-conquer-dp.md

```diff
@@ -41,21 +41,22 @@ template.
 The function `compute` computes one row $i$ of states `dp_cur`, given the previous row $i-1$ of states `dp_before`.
 It has to be called with `compute(0, n-1, 0, n-1)`.
 
-```cpp
+```cpp divide_and_conquer_dp
 int n;
-long long C(int i, int j);
 vector<long long> dp_before(n), dp_cur(n);
 
+long long C(int i, int j);
+
 // compute dp_cur[l], ... dp_cur[r] (inclusive)
-void compute(int l, int r, int optl, int optr)
-{
+void compute(int l, int r, int optl, int optr){
     if (l > r)
         return;
+
     int mid = (l + r) >> 1;
-    pair<long long, int> best = {INF, -1};
+    pair<long long, int> best = {LLONG_MAX, -1};
 
     for (int k = optl; k <= min(mid, optr); k++) {
-        best = min(best, {dp_before[k] + C(k, mid), k});
+        best = min(best, {dp_before[k - 1] + C(k, mid), k});
```

Comment:
It kinda depends a bit on the definitions of `dp_before` and `C` if it is correct or not. E.g. if `dp_before[i]` stores the answer of the array[0...i]` or of `array[0...i-1]`. I usually prefer the second one (but of course then the definition `vector<long long> dp_before(n);` is wrong, and should be `vector<long long> dp_before(n+1);`).

But yeah, we should change it. However your change introduces another bug. What if `k == 0` (which it is the first time you call `compute(0, n-1, 0, n-1)`. You need to handle that special case.

That's also why I usually prefer the other way with `dp_before[i]` being the optimal answer of the `array[0...i-1]`, because then you can just initialize `dp_before[0] = 0` and  don't have and if/else logic inside.

But I'm happy with the change to `k-1`, if you add some special handling of `k-1`. If you do however, we should also change the definition on line 9 in the file, as this one also says `dp(i - 1, k)`. Or we switch to the other method, and increase the dp arrays by one and initialize the first element with 0, and change the line `dp_cur[mid] = ` to `dp_cur[mid+1] = 1`.

Reply:
Not sure why the script ran fine without and all test cases passed. I guess g++ just initialized `dp_before[-1]` with `0` by accident?
Compiling the script with `g++ -std=c++17 -fsanitize=address test_divide_and_conquer_dp.cpp` crashes the program immediately though.

Reply:
Ah, good catch. I guess I trusted the original implementation too much and got careless when the tests passed.

Agreed that it depends on the definitions of `dp_before` and `dp_cur` and how we're making the transitions. Storing the answer for ```array[0...i-1]``` in ```dp[i]``` would make the implementation simpler, but it might be confusing for the readers since all the other ranges are inclusive, but here it will be exclusive.

Therefore feeling inclined to handle the `k=0` case separately as it doesn't degrade the implementation much.

Reply:
Addressed in https://github.com/e-maxx-eng/e-maxx-eng/pull/712/commits/e37adaca646e41010b1a9e0d2bec1e704c072c5b and https://github.com/e-maxx-eng/e-maxx-eng/pull/712/commits/0d2e9f2f67f48ec5b4e591a1804c70062a7e1c58. Please check again.

Reply:
By the way, I got tons of errors after running all the tests with ```-fsanitize=address -fsanitize=undefined``` and a few more after running with some flags mentioned in https://codeforces.com/blog/entry/15547. Perhaps we should fix them and add the flags to the [script](https://github.com/e-maxx-eng/e-maxx-eng/blob/a93dbdb3f33c4f9aa046990d678a6868ee3fead0/test/test.sh#L14) in a separate PR?